std/array_list.zig

```diff
@@ -193,27 +194,105 @@ pub fn AlignedArrayList(comptime T: type, comptime A: u29) type {
             return self.pop();
         }
 
+        // an iterator which allows you to remove the current element
+        // while iterating through it.
         pub const Iterator = struct {
-            list: *const Self,
-            // how many items have we returned
-            count: usize,
+            list: *Self,
+
+            // index of the next element to return
+            next_index: usize = 0,
+
+            // the last index we removed
+            removed_index: ?usize = null,
 
             pub fn next(it: *Iterator) ?T {
-                if (it.count >= it.list.len) return null;
-                const val = it.list.at(it.count);
-                it.count += 1;
+                if (it.next_index >= it.list.len) return null;
+                const val = it.list.at(it.next_index);
+
+                it.next_index += 1;
+
                 return val;
             }
 
             pub fn reset(it: *Iterator) void {
-                it.count = 0;
+                it.next_index = 0;
+                it.removed_index = null;
+            }
+
+            //
+            // NOTE: inlining these removal routines is important.
+            // Allows them to get with 1% of the performance of ArrayList.swapRemove/orderedRemove.
+            //
+
+            pub inline fn swapRemove(it: *Iterator) T {
+                const removed_already = blk: {
+                    if (it.removed_index) |idx| {
+                        break :blk idx==it.next_index;
+                    } else {
+                        break :blk false;
+                    }
+                };
+                if (!removed_already) {
+                    if (it.next_index > 0) it.next_index -= 1;
```

Comment:
Shoudn't this be something like `if (it.next_index == 0) @panic("attempt to remove element before iterating"); it.next_index -= 1;`

Reply:
You can remove before iterating because you're automatically on the first item when you create the iterator. The issue is if there are no items to iterate over but you still try to remove one.

Reply:
Moved this the beginning of the procedure for clarity.

Reply:
If you're automatically "on the first item" then shouldn't `next_item` be 1?
I think you shouldn't be "on an item" until you've called `next`.

Reply:
By "automatically on", I am referring to the next item you'll get from `next`.
You'll automatically get the first item from `next`.

The state is set up so that if you make an iterator and then immediately call for removal, it'll remove the first one in the sequence, or panic if there isn't one.

Reply:
I suggest making it illegal to call a remove function before you call `next`.

Reply:
It's proving a little tricky to do that without any slowdown. But if I can, I will.

Reply:
Scratch that. Figured it out. 😄